exercises/spectre/descr.md

```diff
@@ -0,0 +1,299 @@
+# From a Spectre to a Tree
+
+Suppose we have an algebraic data type `'a tree` of binary trees where each
+leaf carries a piece of data of type `'a`.
+
+```
+  type 'a tree =
+    | Leaf of 'a
+    | Fork of 'a tree * 'a tree
+```
+
+For instance, the tree `Fork (Leaf 1, Fork (Leaf 2, Leaf 3))` has type `int
+tree`, as its leaves carry integers.
+The list of the data values carried by its leaves is `[1; 2; 3]` and the list
+of the depths of its leaves is `[1; 2; 2]`.
+
+In fact, these two lists are sufficient to describe this tree in an
+unambiguous manner: no other three has the same list of leaf data and
+the same list of leaf depths.
+
+Let us combine these lists, and use the word *spectre* to refer a list
+of pairs of a datum and a depth:
+
+```
+  type depth =
+    int
+  type 'a spectre =
+    ('a * depth) list
+```
+
+The spectre of the tree
+`Fork (Leaf 1, Fork (Leaf 2, Leaf 3))`
+is
+`[(1, 1); (2, 2); (3, 2)]`.
+
+A tree is unambiguously described by its spectre: two distinct trees must have
+distinct spectres.
+
+This raises the question: from a spectre, how does one reconstruct a tree?
+
+This simple question gives rise to a nontrivial exercise,
+which can be solved in several ways.
+
+This is really a *parsing* problem. Indeed, parsing consists in reconstructing
+a tree out of a purely sequential representation of this tree. Furthermore,
+because the spectre of a tree carries the depth of each leaf, reconstructing
+a tree from a spectre is analogous to *indentation-aware parsing*.
+
+## Warmup: From a Tree to its Spectre
+
+**Question 1.** Define a function `elements` of type `'a tree -> 'a list`
+that computes the list of the data values carried by the leaves of a tree.
+
+**Question 2.** Define a function `depths` of type `'a tree -> depth list`
+that computes the list of the depths of the leaves of a tree.
+
+**Question 3.** Define a function `spectre` of type `'a tree -> 'a spectre`
```

Comment:
I guess the orders of appearance of the elements and of the depths do not matter for the grader as long as they are consistent with each other.

Reply:
I have second thoughts about this: you are actually assuming a list obtained by a left-to-right traversal of the leaves, aren't you? Otherwise, I guess that it is not possible to correctly recover the original tree by left-to-right parsing.

Reply:
The functions `elements`, `depths` and `spectre` must read the leaves of the tree from left to right. The order of the elements in the lists that are returned matters. I have tried to clarify this at the beginning.